integration/cursors/awaitdata_test.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cursors
+
+import (
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+
+	"github.com/FerretDB/FerretDB/integration"
+	"github.com/FerretDB/FerretDB/integration/setup"
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/testutil"
+)
+
+func TestCursorsTailableAwaitData(tt *testing.T) {
+	tt.Parallel()
+
+	t := setup.FailsForFerretDB(tt, "https://github.com/FerretDB/FerretDB/issues/2283")
+
+	s := setup.SetupWithOpts(t, nil)
+
+	db, ctx := s.Collection.Database(), s.Ctx
+
+	opts := options.CreateCollection().SetCapped(true).SetSizeInBytes(10000000000)
+	err := db.CreateCollection(s.Ctx, testutil.CollectionName(t), opts)
+	require.NoError(t, err)
+
+	collection := db.Collection(testutil.CollectionName(t))
+
+	_, err = collection.InsertOne(ctx, bson.D{{"v", "foo"}})
+	require.NoError(t, err)
+
+	cmd := bson.D{
+		{"find", collection.Name()},
+		{"batchSize", 1},
+		{"tailable", true},
+		{"awaitData", true},
+		{"maxTimeMS", 1},
+	}
+
+	var res bson.D
+	err = collection.Database().RunCommand(ctx, cmd).Decode(&res)
+	require.NoError(t, err)
+
+	var firstBatch *types.Array
+	firstBatch, cursorID := getFirstBatch(t, res)
+
+	require.Equal(t, 1, firstBatch.Len())
+
+	getMoreCmd := bson.D{
+		{"getMore", cursorID},
+		{"collection", collection.Name()},
+		{"batchSize", 1},
+		{"maxTimeMS", (10 * time.Minute).Milliseconds()},
+	}
+
+	insertChan := make(chan error)
+
+	go func() {
+		time.Sleep(1 * time.Second)
+		_, insertErr := collection.InsertOne(ctx, bson.D{{"v", "bar"}})
+		insertChan <- insertErr
+	}()
+
+	err = collection.Database().RunCommand(ctx, getMoreCmd).Decode(&res)
+	require.NoError(t, err)
+
+	require.NoError(t, <-insertChan)
+
+	nextBatch, nextID := getNextBatch(t, res)
+	require.Equal(t, cursorID, nextID)
+	require.Equal(t, 1, nextBatch.Len())
```

Comment:
Ok, now I don't understand why we use a goroutine at all. There is no blocking anywhere. `res` is not checked after `RunCommand` and before waiting for `insert` to complete. `getNextBatch` doesn't do any I/O. What are we testing there, exactly?

Reply:
What's the difference between checking `res` before or after getting insertChan content? 

In this test we want to:
1) Run `getMore` command on tailable awaitData cursor, that has 10 minutes of maxTime.
2) After some time (1 second) insert data in parallel, so the previously called getMore command unblock and fetch the new document
3) validate that `getMore` batch has 1 document. If it had 0 it'd mean that cursor didn't block which is incorrect behavior